### ISSUE TITLE
Polymorphic `Fail` effect?

### ISSUE THREAD

**Issue by mitchellwrosen** [fused-effects/fused-effects#154]
What do you think about making the `Fail` effect take a type variable rather than being hard-coded to `String`? (The `MonadFail` instance can be recovered with a `a ~ String` constraint).

**Comment by robrix** [comment]
@mitchellwrosen: Could you please describe your motivating use case(s) for `Fail a` where `a` is not `String`? I’m especially curious about anything which would make `Error a` undesirable.

**Comment by mitchellwrosen** [comment]
Sure! I have a `servant` app, and the handlers may be written in any monad you want as long as you provide a natrual transformation `m ~> ExceptT ServerError IO`.

I have a handler that may throw a domain-specific error, say `FooErr`. My first thought was to use the `Error` effect:

```haskell
fooHandler :: (Carrier sig m, Member (Error FooErr) => m Text
```

Now one option for my `m` monad is to include an `ErrorC FooErr` layer and end up with an `Either FooErr a` then convert that to a `ServerError` in my natural transformation at the top level.

But I thought it would be cleaner (conceptually) to use a fused-effect carrier to interpret a `FooErr` _as_ a `ServerError` by simply printing it, then converting it to a generic `500` response.

So to recap, my goal was to "rewrite"

```
Member (Error FooErr) m
```

to

```
(Member (Error ServerError) m, MonadIO m)
```

(MonadIO for printing).

Since I want to get rid of `FooErr` entirely I thought the appropriate carrier would simply be a newtype around `m a` -

```haskell
newtype RewriteFooErrC m a = RewriteFooErrC m a
```

and its `Carrier` instance would require `m` to be able to throw `ServerError`s. However, I got stuck implementing this, because it's not possible to handle `Catch` nodes in a law-abiding way. Since my interpretation `m a` contains no information about whether a `FooErr` was thrown, I can't pass any such error to the exception handler contained within.

**Comment by mitchellwrosen** [comment]
I thought it might be possible to rewrite/unify errors this way using an effect that only had a `Throw` node, not `Catch`. Then the strategy would be:

- Stay "open" in code that has a typed error it may throw by using the `Error` effect (since someone upstream may want to catch it).

- At the layer of the app where we're sure we don't want to `catchError` anything, rewrite `Error` effects to a `Fail`-like effect that only has a `Throw` node (via a simple carrier).

- Then, unify `Fail` effects as necessary, again using carriers.

**Comment by mitchellwrosen** [comment]
Well I just spent some time trying to implement the above and the machinery did not work out how I envisioned... so maybe I don't have a need for a polymorphic `Fail` after all.

**Comment by patrickt** [comment]
I would be reluctant to include a `Throw` effect in fused-effects core, but I think this is an excellent fix for a third-party package, given the reasonable justification above. 

If I’m understanding your initial request, @mitchellwrosen, would this function help?
```
import Data.Bifunctor (first)

rehandling :: (w -> e) -> ErrorC w m a -> ErrorC e m a
rehandling h (ErrorC f) = ErrorC (first h f)
```

**Comment by robrix** [comment]
I mostly agree with @patrickt; in particular, I think one of the virtues of the `Fail` effect is its simplicity and suitability to a single purpose.

However, I do think it would be interesting, and perhaps useful enough to warrant inclusion, to define `Throw` and `Catch` effects, each providing only one of the `Error` constructors. `ErrorC` would then handle both, similarly to how `CutC` and `CullC` handle `NonDet` effects in addition to `Cut` & `Cull` respectively.

Meanwhile, I believe @mitchellwrosen’s case would still be accommodated by rethrowing anything not caught into an underlying `Throw` effect (again, quite like `CullC`).

I still have some reluctance to include this in `fused-effects` proper, however, centred on doubts as to whether it’ll be sufficiently approachable & convenient for users to see (and be expected to use) separate `Member (Throw err) sig` and `Member (Catch err) sig` constraints, with nothing obviously relating them except the type of `err` itself.

I think I’ll have to play with this some to decide whether it should be in the library. In the meantime I’d love to hear your experience trying this in your app, @mitchellwrosen!

**Comment by mitchellwrosen** [comment]
@patrickt Close, I don't quite want `rehandling`, because I want to translate errors of one type to another while performing other effects (like logging).

@robrix Excellent idea to have two interacting effects for `Throw` and `Catch`! I completely forgot about that feature of fused-effects. That would solve my issue entirely.

Here is a snippet of fused-effects working really well (rewriting high-level domain-specific effects to lower and lower levels)

```haskell
          runGitHubAuthHttp gitHubClientId gitHubClientSecret
      >>> runHttpRequestIO httpManager

          -- Persistence layer
      >>> runPersistPollDB
      >>> runPersistPollAnswerDB
      >>> runPersistUserDB
      >>> runDBC postgresPool

          -- HTTP session
      >>> runHttpSessionIO cookieSettings jwtSettings

          -- Event handlers
      >>> runYieldChan (unsafeTBroadcastChanToTChan pollAnsweredEventChan)
      >>> runYieldChan (unsafeTBroadcastChanToTChan pollCreatedEventChan)
```

but then error handling got awkward (due to having to rewrite `Error` to `Fail'`, a polymorphic version of `Fail`)

```haskell
          -- Error handlers
      >>> runErrorFail @Hasql.UsageError
      >>> runHasqlUsageErrorLogAnd500
      >>> runErrorFail @Servant.ClientError
      >>> runServantClientErrorLogAnd500
      >>> runError @Servant.ServerError
```

but separate `Throw`/`Catch` effects will clean that up. Thanks!

**Comment by robrix** [comment]
@mitchellwrosen: Cool, thanks for circling back with your experience! We’ll definitely consider it—I agree that `fail` is a bit of an anachronism, and I personally only use it as a stepping stone to some “real” error type.

**Comment by robrix** [comment]
I want to split `Error` up into separate `Catch` & `Throw` effects, with `Error` still provided as a type synonym for `Catch :+: Throw`. This is going to require a solution to #213, which itself depends on some compilation time measurements that I haven’t taken yet. But I wanted to take a moment to write down my rationale:

Splitting `NonDet` up into `Choose` & `Empty` effects in #199 allowed us to express both failure-without-branching and branching-without-failure, both useful models of computation in their own right. `Throw` is an obvious generalization of both `Empty` and `Fail` (as noted here), and furthermore has the potential for interesting interactions with `Choose`, e.g. nondeterminism with both branching and failure, where failures can carry arbitrary error information. But you might not want both that _and_ the ability to catch such errors—essentially another type of nondeterminism—thrown into the mix. For example, what should this mean?

```haskell
(throwError err <|> k) `catchError` f
```

There are multiple reasonable interpretations of this, and that’s valuable—but you might not want to surface this to users of your API.

On the other hand, you might want to combine `Empty` and `Catch` such that users can’t throw arbitrary errors, merely abort a computation, but the underlying carrier can capture its internal state at the time the abort occurred and supply that to the error handler.

These aren’t academic, either—both of these have come up when writing parsing systems such as `semantic`’s `Assignment` DSL.

So, `Throw` is useful on its own, and in combination with other effects (including `Catch`); `Catch` is useful in combination with other effects (including `Throw`); all in all, this gives me reason to believe that we should split them up—provided that we can still maintain our ergonomics (again, cf #213).

**Comment by robrix** [comment]
@mitchellwrosen: We just merged #247 in, hope you enjoy the new flexibility!
